text/0000-module-resolution.md

```diff
@@ -0,0 +1,340 @@
+  - Tools like Webpack or RollupJs allow you to configure and customize the way we load bundle and resolve modules both run build-time and run-time.
+
+## Why we need our own module resolution
+
+Web Components are composed of multiple pieces (CSS, HTML and JS) and there is no canonical or standard way to bundle them natively yet (altough there are several proposals to standarize the way by which each of this pieces can interoperate and be integrated in the future see [CSS modules](https://github.com/w3c/webcomponents/issues/759), [HTML templates](https://github.com/w3c/webcomponents/issues/839)) hence we must build a future-proof, standard compliant way to do so while we help the standarization pieces to land.
+
+Even when we have all this pieces become available (and without going into too much detail here), it is likely that we will have to do some preprocessing or bundling anyway, which mean that we will have to take over the resolution one way or another - wether is build time to collapse mutliple pieces in on, or at runtime by using import-maps with some extra semantics on top. Moreover remember that a give package might contain many components (1:N) relationship, so will always have to have a configuration on where to find those components.
+
+To summarize: We will always need an abstraction layer which dictates where a particular module specifier must be resolved from (adapted to different host environments).
+
+## Detailed design
+
+In order to resolve modules in LWC that work on different host environments (Salesforce platform, Node.js, npm, git, et all) we will be relying on our own LWC configuration file (similar approach as import-maps, babel, webpack rollup or package.json).
+
+### LWC configuration
+
+A configuration file will be located at the root of the project, optionally at the component namespace level and in the future (this is yet to be spec out) at the module bundle level, where specificitiy will be resolved from inner to outer (merging upwards - similar to node_modules).
+
+At the package root level there are two ways on which you can add a configuration file: Creating a `lwc.config.json` or adding an `lwc` key to package.json (this is mostly for simplicity and easy of use - specially when publishing a package). Here are some examples:
+
+```json5
+// package.json
+{
+    "name: "foo",
+    "version": "bar",
+    "lwc": {
+        "modules": [
+            /* list with modules records goes here */
+        ]
+    }
+}
+```
+
+```json5
+// lwc.config.json
+{
+    "modules:" [
+        /* list with modules records goes here */
+    ]
+}
+```
+
+Its important that **any plugin, tool or project honor this configuration**, otherwise the resolution will not work correctly in most environments.
+
+Moreover, note that this type of configuraiton is not something new, many tools use an equivalent configuration (babel, lerna, rollup, webpack, ...)
```

Comment:
Great point. It was feeling kind of weird to have to have this explicitly defined but everyone does it.